cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,977 @@
+
+Most algorithms must support columns of any data type. This leads to complexity in the code, and 
+is one of the primary challenges a libcudf developer faces. Sometimes we develop new algorithms with
+gradual support for more data types to make this easier. Typically we start with fixed-width data
+types such as numeric types and timestamps/durations, adding support for nested types later.
+
+Enabling an algorithm differently for different types uses either template specialization or SFINAE,
+as discussed in [Specializing Type-Dispatched Code Paths](#specializing-type-dispatched-code-paths).
+
+# Type Dispatcher
+
+libcudf stores data (for columns and scalars) "type erased" in `void*` device memory. This 
+*type-erasure* enables interoperability with other languages and type systems, such as Python and 
+Java. In order to determine the type, libcudf algorithms must use the run-time information stored 
+in the column `type()` to reconstruct the data type `T` by casting the `void*` to the appropriate 
+`T*`.
+
+This so-called *type dispatch* is pervasive throughout libcudf. The `type_dispatcher` is a 
+central utility that automates the process of mapping the runtime type information in `data_type` 
+to a concrete C++ type.
+
+At a high level, you call the `type_dispatcher` with a `data_type` and a function object (also 
+known as a *functor*) with an `operator()` template. Based on the value of `data_type::id()`, the 
+type dispatcher invokes the corresponding instantiation of the `operator()` template. 
+
+This simplified example shows how the value of `data_type::id()` determines which instantiation of 
+the `F::operator()` template is invoked.
+
+```c++
+template <typename F>
+void type_dispatcher(data_type t, F f){
+    switch(t.id())
+       case type_id::INT32: f.template operator()<int32_t>()
+       case type_id::INT64: f.template operator()<int64_t>()
+       case type_id::FLOAT: f.template operator()<float>()
+       ...
+}
+```
+
+The following example shows a function object called `size_of_functor` that returns the size of the 
+dispatched type.
+
+```c++
+struct size_of_functor{
+  template <typename T>
+  int operator()(){ return sizeof(T); }
+};
+
+cudf::type_dispatcher(data_type{type_id::INT8}, size_of_functor{});  // returns 1
+cudf::type_dispatcher(data_type{type_id::INT32}, size_of_functor{});  // returns 4
+cudf::type_dispatcher(data_type{type_id::FLOAT64}, size_of_functor{});  // returns 8
+```
+
+By default, `type_dispatcher` uses `cudf::type_to_id<t>` to provide the mapping of `cudf::type_id` 
+to dispatched C++ types. However, this mapping may be customized by explicitly specifying a 
+user-defined trait for the `IdTypeMap`. For example, to always dispatch `int32_t` for all values of 
+`cudf::type_id`:
+
+```c++
+template<cudf::type_id t> struct always_int{ using type = int32_t; }
+
+// This will always invoke `operator()<int32_t>`
+cudf::type_dispatcher<always_int>(data_type, f);
+```
+
```

Comment:
I feel we need a Spider-Man quote here: _With great power comes great responsibility._

Try to avoid multiple type-dispatch if possible. The compiler will create a code path for every type so a 2nd type-dispatch would multiply the paths by the number of types. This can create very large object code and cause very long compile times.

Reply:
Added